File: strategy/marvin/task/aggressivekeeper.lua
local ForceShoot = require "task/ability/forceshoot"
local AggressiveKeeper = Class("Task.AggressiveKeeper",
	require "task/base", ForceShoot)

local Field = require "../base/field"
local geom = require "../base/geom"
local vis = require "../base/vis"
local World = require "../base/world"
local Physics = require "observer/physics"
local Robot = require "observer/robot"
local PathHelper = require "trajectory/pathhelper"
local ToTarget = require "trajectory/totarget"


local POSITION_PADDING = 0.02 -- safety distance

local CHIP_IMPACT_DIST_FROM_BORDER = 0.5
local CHIP_DIST_FACTOR = 0.25
local CHIP_GOAL_LINE_DIST = 1

function AggressiveKeeper:run()
	local safeGoalMid = World.Geometry.FriendlyGoal - Vector(0, 0.05)
	local moveDest
	local ignoreBall
	if World.Ball.pos.y < self._robot.pos.y + POSITION_PADDING then
		-- get between ball and goal
		local ballDist = self._robot.radius + World.Ball.radius
		moveDest = World.Ball.pos + (safeGoalMid - World.Ball.pos):setLength(ballDist) + Vector(0, -POSITION_PADDING)
		ignoreBall = false
	else
		local ballTime = Robot.minTimeToBall(self._robot)
		moveDest = Physics.ballAtTime(World.Ball, ballTime).pos
		moveDest = moveDest + (self._robot.pos - moveDest):setLength(World.Ball.radius)
		ignoreBall = true
	end

	self:_chipToBorderIfSafe()

	PathHelper.setDefaultObstacles(self._robot.path, self._robot, ignoreBall)
	local viewDir = World.Ball.pos - safeGoalMid
	self._robot.trajectory:update(ToTarget, moveDest, viewDir:angle(), nil, viewDir * 0.5)
end


local touchLineDir = Vector(0, 1)

local leftFriendlyCorner = Vector(-World.Geometry.FieldWidthHalf, -World.Geometry.FieldHeightHalf)
local rightFriendlyCorner = Vector(World.Geometry.FieldWidthHalf, -World.Geometry.FieldHeightHalf)

-- assume chips crossing this line might cross the goal line
local leftNearBasePoint = Vector(-World.Geometry.FieldWidthHalf, World.Geometry.FieldHeightHalf-CHIP_GOAL_LINE_DIST)
local rightNearBasePoint = Vector(World.Geometry.FieldWidthHalf, World.Geometry.FieldHeightHalf-CHIP_GOAL_LINE_DIST)
local nearBaseLineDir = rightNearBasePoint-leftNearBasePoint

function AggressiveKeeper:_chipToBorderIfSafe()
	local robotPos = self._robot.pos
	local ballPos = World.Ball.pos
	local robotDir = ballPos - robotPos
	local viewAngle = robotDir:angle()
	local rigthCornerAngle = (rightFriendlyCorner - robotPos):angle()
	local leftCornerAngle = (leftFriendlyCorner - robotPos):angle()
	if viewAngle > rigthCornerAngle or viewAngle < leftCornerAngle then -- not towards own goal line
		local touchLineIntersection = Field.nextLineCut(robotPos, robotDir)
		local chipPos = geom.intersectLineLine(robotPos, robotDir, leftNearBasePoint, nearBaseLineDir)

		if chipPos and touchLineIntersection then
			if robotPos:distanceTo(touchLineIntersection) < robotPos:distanceTo(chipPos) then
				chipPos = touchLineIntersection
			end
		elseif touchLineIntersection then -- no nearBaseline
			chipPos = touchLineIntersection
		else -- probably because ball is out of field
			chipPos = World.Geometry.OpponentGoal
		end
		local chipDist = World.Ball.pos:distanceTo(chipPos) - CHIP_IMPACT_DIST_FROM_BORDER
		if chipPos ~= touchLineIntersection then -- try to avoid icing if chipping towards the opponent goal line
			chipDist = chipDist*CHIP_DIST_FACTOR
		end

		vis.addCircle("t/a/chipToBorder", ballPos + robotDir:copy():setLength(chipDist), 0.1, vis.colors.blue, true)
		if not Robot.hadBall(self._robot, 0) then
			self._forceShootTimer = nil
		end
		self:_doForceShoot()
		self._robot:chip(chipDist)
	end
end

return AggressiveKeeper

File: strategy/typescript/glados/task/keeper/aggressivekeeper.ts
import * as Field from "base/field";
import * as geom from "base/geom";
import { Vector } from "base/vector";
import * as vis from "base/vis";
import * as World from "base/world";

import * as Physics from "glados/observer/physics";
import * as Robot from "glados/observer/robot";
import { ForceShoot } from "glados/task/ability/forceshoot";
import { Agent, Task } from "glados/task/base";
import { CurvedMaxAccel } from "glados/trajectory/curvedmaxaccel";
import * as PathHelper from "glados/trajectory/pathhelper";


const POSITION_PADDING = 0.02; // safety distance

const CHIP_IMPACT_DIST_FROM_BORDER = 0.5;
const CHIP_DIST_FACTOR = 0.25;
const CHIP_GOAL_LINE_DIST = 1;

const leftFriendlyCorner = new Vector(-World.Geometry.FieldWidthHalf, -World.Geometry.FieldHeightHalf);
const rightFriendlyCorner = new Vector(World.Geometry.FieldWidthHalf, -World.Geometry.FieldHeightHalf);

// assume chips crossing this line might cross the goal line
let leftNearBasePoint = new Vector(-World.Geometry.FieldWidthHalf, World.Geometry.FieldHeightHalf - CHIP_GOAL_LINE_DIST);
let rightNearBasePoint = new Vector(World.Geometry.FieldWidthHalf, World.Geometry.FieldHeightHalf - CHIP_GOAL_LINE_DIST);
let nearBaseLineDir = rightNearBasePoint - leftNearBasePoint;

export class AggressiveKeeper extends Task {
	private _forceShoot: ForceShoot;

	constructor(agent: Agent) {
		super(agent);
		this._forceShoot = new ForceShoot(this._robot);
	}

	run() {
		const MAX_DIST_TO_DEFENSE_AREA = 0.3;

		let safeGoalMid = World.Geometry.FriendlyGoal - new Vector(0, 0.05);
		let moveDest;
		let ignoreBall;
		if (World.Ball.pos.y < this._robot.pos.y + POSITION_PADDING) {
			// get between ball and goal
			let ballDist = this._robot.radius + World.Ball.radius;
			moveDest = World.Ball.pos + (safeGoalMid - World.Ball.pos).setLength(ballDist) + new Vector(0, -POSITION_PADDING);
			ignoreBall = false;
		} else {
			let ballTime = Robot.minTimeToBall(this._robot);
			moveDest = Physics.ballAtTime(World.Ball, ballTime).pos;
			moveDest = moveDest + (this._robot.pos - moveDest).setLength(World.Ball.radius);
			ignoreBall = true;
		}

		this._chipToBorderIfSafe();

		let ignoreFriendlyRobots = false;
		if (Field.distanceToFriendlyDefenseArea(this._robot.pos, this._robot.radius) < MAX_DIST_TO_DEFENSE_AREA) {
			ignoreFriendlyRobots = true;
		}
		let obstacleTable = {
			ignoreBall: ignoreBall,
			ignorePass: true,
			ignoreFriendlyRobots: ignoreFriendlyRobots,
			disableOpponentPrediction: true,
			messaging: this._messaging
		};
		PathHelper.setDefaultObstaclesByTable(this._robot.path, this._robot, obstacleTable);
		let viewDir = World.Ball.pos - safeGoalMid;
		this._robot.trajectory.update(CurvedMaxAccel, moveDest, viewDir.angle(), undefined, viewDir * 0.5);
	}

	_chipToBorderIfSafe() {
		let robotPos = this._robot.pos;
		let ballPos = World.Ball.pos;
		let robotDir = ballPos - robotPos;
		let viewAngle = robotDir.angle();
		let rigthCornerAngle = (rightFriendlyCorner - robotPos).angle();
		let leftCornerAngle = (leftFriendlyCorner - robotPos).angle();
		if (viewAngle > rigthCornerAngle || viewAngle < leftCornerAngle) { // not towards own goal line
			let touchLineIntersection = Field.nextLineCut(robotPos, robotDir);
			let chipPos = geom.intersectLineLine(robotPos, robotDir, leftNearBasePoint, nearBaseLineDir)[0];

			if (chipPos != undefined && touchLineIntersection != undefined) {
				if (robotPos.distanceTo(touchLineIntersection) < robotPos.distanceTo(chipPos)) {
					chipPos = touchLineIntersection;
				}
			} else if (touchLineIntersection != undefined) { // no nearBaseline
				chipPos = touchLineIntersection;
			} else {// probably because ball is out of field
				chipPos = World.Geometry.OpponentGoal;
			}
			let chipDist = World.Ball.pos.distanceTo(chipPos) - CHIP_IMPACT_DIST_FROM_BORDER;
			if (chipPos !== touchLineIntersection) { // try to avoid icing if chipping towards the opponent goal line
				chipDist = chipDist * CHIP_DIST_FACTOR;
			}

			vis.addCircle("t/a/chipToBorder", ballPos + robotDir.copy().setLength(chipDist), 0.1, vis.colors.blue, true);
			if (!Robot.hadBall(this._robot, 0)) {
				this._forceShoot._forceShootTimer = undefined;
			}
			this._forceShoot._doForceShoot();
			this._robot.chip(chipDist);
		}
	}
}
